### ISSUE TITLE
Dictionary fails to parse JSON with array as top level

### ISSUE THREAD

**Issue by neikeq** [godotengine/godot#4232]
**Operating system or device:**
Ubuntu 15.10 64bits

**Issue description** (what happened, and what was expected):
It's valid for a JSON to have an array as top level, however _Dictionary_ fails to parse it:

```
ERROR: parse_json: Error parsing JSON: Expected '{' at line: 0
   At: core/dictionary.cpp:206.
```

**Steps to reproduce:**

``` GDScript
func test_json_top_level_array():
    assert({}.parse_json("[ { \"name\": \"Richard\" }, {\"name\": \"Drahcir\"} ]") == OK)
```

**Comment by blurymind** [comment]
This bug affects me :)

**Comment by neikeq** [comment]
@reduz The problem is that the root of a Dictionary needs at least one key. IMO the best solution would be to move expose a JSON class to GDScript with a parse method that would return either an Array or a Dictionary.

**Comment by mooglegiant** [comment]
top level still needs { } and dictionaries/arrays need a key
it should be something like
{ \"objecttype\" : [ { \"name\": \"Richard\" }, {\"name\": \"Drahcir\"} ] }

when it doubt, validate with http://json.parser.online.fr/ (just remove the escape characters)

**Comment by neikeq** [comment]
@mooglegiant A valid JSON can have an array [] as top level.

**Comment by Razzlegames** [comment]
But valid Json doesn't necessarily translate toba valid dictionary. Json that maps to a dictionary is a subset of all possible JSON

**Comment by bojidar-bg** [comment]
@Razzlegames ~ I think you didn't actually read @neikeq's proposal above, which is to move JSON parsing to a new `JSON` class, and out of `Dictionary` (though the latter should be kept for compatibility). That would allow the `parse_json` method to return an `Array` as well as a `Dictionary`.

**Comment by neikeq** [comment]
@bojidar-bg I agree. Dictionary could still keep `parse_json` with the current behaviour.

**Comment by punto-** [comment]
It could also be on Marshalls, which is where all the serialization stuff
is.

**Comment by bojidar-bg** [comment]
@punto- I think rather not, since in Marshalls most stuff is related to Variant-RawArray conversion, and not JSON or some other human-readable format.

**Comment by blurymind** [comment]
This will help me port yarn parser to godot. The jason files yarn Exports
have [] as root

**Comment by vnen** [comment]
I never notice before, but it seems that `str2var()` can understand JSON, be it dictionary or array.

**Comment by neikeq** [comment]
@vnen That's probably because a Dictionary as string has a similar syntax to JSON. But is it exactly the same in all cases?

**Comment by vnen** [comment]
I don't know. It certainly manages simple JSON, but not sure how much of it (haven't tested things like escape sequences or `true`, `false`, `null` values).

**Comment by reduz** [comment]
I've been re-reading the JSON specification, apparently anything can be at
top level in JSON, even strings, ints, etc.
So, adding this ability to Array is not a good idea. It should be added to
String and return a Variant, however due to how C++ works this is not
possible (cross dependency).

So, if anyone is up for fixing this, the plan should be:

For C++ we should just use the JSON singleton instead of Dictionary (make a
static function like Variant JSON::parse(const String&) ), and for GDScript
we can add a binding manually to String in variant_call.cpp

We can keep the current function in Dictionary for compatibility and erase
it in 3.0

**Comment by reduz** [comment]
currently, the JSON::parse function expects a { token, as it calls
parse_object(). It should be changed to use _parse_value()

**Comment by neikeq** [comment]
OK, I tried validating `"foo"` with https://jsonformatter.curiousconcept.com/
```
Invalid JSON (RFC 4627)
Valid JSON (RFC 7159)
Invalid JSON (ECMA-404)
```
The last two are the competing standards. Would supporting `RFC 7159` break compatibility with `ECMA-404`?

BTW, I can make the PR.

**Comment by vnen** [comment]
According to Ecma 404:

> A JSON text is a sequence of tokens formed from Unicode code points that conforms to the JSON value grammar.

and in the value grammar:

> A JSON value can be an object, array, number, string, true, false, or null. 

So I'd say it actually agrees with RFC 7159 ans that JSON formatter tool misinterpreted it. Also, this RFC states:

> Implementations that generate only objects or arrays where a JSON text is called for will be interoperable in the sense that all implementations will accept these as conforming JSON texts.

If we try to parse and it's an array or object, it'll be valid. The major concern would be about JSON *generartion*, but I think we can assume the user will pass a valid dictionary/array if they want such format (and this will be documented as such).

Also, by both specs (as I understand) `"foo","bar"` is valid JSON. The formatter says this is valid according to RFC 7159. This might be interpreted as an array for our purposes.

**Comment by reduz** [comment]
I don't think I would go that far, let's just expect arrays to have [] but
accept any token such as "helo" or 25 as valid JSON

**Comment by Razzlegames** [comment]
As long as it's documented, I don't think there's any problem with only
supporting a subset of the JSON standard.

**Comment by reduz** [comment]
I wrote the parser based on this spec:
http://www.json.org/
though I clearly misunderstood that everything had to be inside of a dictionary because 99.999999% of JSON is used like that :P

over there, it doesn't show that you can do 1,2,3 and treat it as array

**Comment by akien-mga** [comment]
Is this fixed with 62273e5?

**Comment by neikeq** [comment]
Tested on 62273e5 with the following code:

``` GDScript
extends Node

var json_list = [
	"0",
	"\"hello\"",
	"{\"hello\":10}",
	"[\"hello\", 10]",
	"[{\"hello\": 10}]"
]

func _ready():
	for json in json_list:
		assert(validate_json(json).empty())
		var parsed = parse_json(json)
		print(parsed)
		var back = to_json(parsed)
		print(back)

```

Output:

```
0
0
hello
"hello"
(hello:10)
{"hello":10}
[hello, 10]
["hello", 10]
[(hello:10)]
[{"hello":10}]
```

It works!
